ggml-qnn.cpp

```diff
@@ -0,0 +1,3588 @@
+    set_qnn_tensor_datatype(*tensor, dataType);
+}
+
+
+static inline void set_qnn_tensor_quantparams(Qnn_Tensor_t & tensor, Qnn_QuantizeParams_t params) {
+    if (tensor.version == QNN_TENSOR_VERSION_1) {
+        tensor.v1.quantizeParams = params;
+    }
+}
+
+
+[[maybe_unused]] static inline void set_qnn_tensor_quantparams(Qnn_Tensor_t * tensor, Qnn_QuantizeParams_t params) {
+    set_qnn_tensor_quantparams(*tensor, params);
+}
+
+
+static inline void set_qnn_tensor_rank(Qnn_Tensor_t & tensor, uint32_t rank) {
+    if (tensor.version == QNN_TENSOR_VERSION_1) {
+        tensor.v1.rank = rank;
+    }
+}
+
+
+[[maybe_unused]] static inline void set_qnn_tensor_rank(Qnn_Tensor_t * tensor, uint32_t rank) {
+    set_qnn_tensor_rank(*tensor, rank);
+}
+
+
+static inline void set_qnn_tensor_dimensions(Qnn_Tensor_t & tensor, uint32_t * dims) {
+    if (tensor.version == QNN_TENSOR_VERSION_1) {
+        tensor.v1.dimensions = dims;
+    }
+}
+
+
+[[maybe_unused]] static inline void set_qnn_tensor_dimensions(Qnn_Tensor_t * tensor, uint32_t * dims) {
+    set_qnn_tensor_dimensions(*tensor, dims);
+}
+
+
+static inline void set_qnn_tensor_memtype(Qnn_Tensor_t & tensor, Qnn_TensorMemType_t memType) {
+    if (tensor.version == QNN_TENSOR_VERSION_1) {
+        tensor.v1.memType = memType;
+    }
+}
+
+
+[[maybe_unused]] static inline void set_qnn_tensor_memtype(Qnn_Tensor_t * tensor, Qnn_TensorMemType_t memType) {
+    set_qnn_tensor_memtype(*tensor, memType);
+}
+
+
+static inline void set_qnn_tensor_clientbuf(Qnn_Tensor_t & tensor, Qnn_ClientBuffer_t clientBuf) {
+    if (tensor.version == QNN_TENSOR_VERSION_1) {
+        tensor.v1.clientBuf = clientBuf;
+    }
+}
+
+
+[[maybe_unused]] static inline void set_qnn_tensor_clientbuf(Qnn_Tensor_t * tensor, Qnn_ClientBuffer_t clientBuf) {
+    set_qnn_tensor_clientbuf(*tensor, clientBuf);
+}
+
+
+static inline void set_qnn_tensor_memhandle(Qnn_Tensor_t & tensor, Qnn_MemHandle_t handle) {
+    if (tensor.version == QNN_TENSOR_VERSION_1) {
+        tensor.v1.memHandle = handle;
+    }
+}
+
+
+[[maybe_unused]] static inline void set_qnn_tensor_memhandle(Qnn_Tensor_t * tensor, Qnn_MemHandle_t handle) {
+    set_qnn_tensor_memhandle(*tensor, handle);
+}
+
+
+static size_t memscpy(void * dst, size_t dstSize, const void * src, size_t copySize) {
+    if (!dst || !src || !dstSize || !copySize)
+        return 0;
+
+    size_t minSize = dstSize < copySize ? dstSize : copySize;
+
+    memcpy(dst, src, minSize);
+
+    return minSize;
+}
+
+
+static char * ggml_qnn_strndup(const char * source, size_t maxlen) {
+    return ::strndup(source, maxlen);
+}
+
+
+static int deep_copy_qnn_tensors(Qnn_Tensor_t & src, Qnn_Tensor_t & dst) {
+    int err = 0;
+    VALIDATE_TENSOR_VERSION(src, err);
+
+    dst.version = src.version;
+    QNN_TENSOR_SET_NAME(
+            dst, ggml_qnn_strndup(QNN_TENSOR_GET_NAME(src), std::string(QNN_TENSOR_GET_NAME(src)).size()));
```

Comment:
looks the memory allocated by `strndup` will never get free

reference: [strndup](https://en.cppreference.com/w/c/string/byte/strndup)